File: Depth_First_Search/Depth_First_Search.dart
// Required library to use inbuilt queue in LIFO manner in Dart
import 'dart:collection';
// Required library to use standard output facility of dart
import 'dart:io';

// Function to find the depth first search of a graph
void Depth_First_Search ( var graph, int source, int no_vertices ) {
    // Queue is used in LIFO manner to obtain stack functionality.
    var queue = Queue();
    // A list to keep track of visited nodes
    var visited = new List(no_vertices);
    // Initializing the list
    for(int i = 0; i < no_vertices; i ++ ) {
        visited[i] = 0;
    }
    // Adding the source node to LIFO queue ( stack )
    queue.add(source);
    // Marking the node visited
    visited[source] = 1;
    int val;

    while(queue.isNotEmpty){
        val = queue.removeLast();
        stdout.write("$val ");
        
        //Visiting neighbours of the popped node  
        for(int i = 0; i < no_vertices; i ++ ){
            if(graph[val][i] == 1 && visited[i] == 0){
                queue.add(i);
                visited[i] = 1;
            }
        }
    } 
}

// Driver method of the program
void main(){

    // Reading from the standard input
    print("Which graph do you want to enter ( directed/ undirected ):");
    String mode = stdin.readLineSync().toString();

    print("Enter number of vertices:");
    var input = stdin.readLineSync();
    int no_vertices = int.parse(input);
    
    // Graph is implemented as adjacency matrix.
    var graph = new List.generate(no_vertices, (_) => new List(no_vertices));
    for(int i = 0; i < no_vertices; i++){
        for(int j = 0; j < no_vertices; j++){
            graph[i][j] = 0;
        }
    }
  
    print("Enter number of edges:");
    var input1 = stdin.readLineSync();
    int no_edges = int.parse(input);
    print("Enter source and destination vertices of edges:\n");
    print("Note: Enter source and destination vertices as 0,1,2,...");
    
    for(int i = 1; i <= no_edges; i++){
      
        print("Edge #$i:");
        
        print("Enter source vertex:");
        int source = int.parse(stdin.readLineSync());

        print("Enter destination vertex:");
        int destination = int.parse(stdin.readLineSync());

        if(mode == "undirected"){
            graph[source][destination] = 1;
            graph[destination][source] = 1;
        }
        else{
            graph[source][destination] = 1;
        }

    }

    print("Enter the source vertex from which DFS should begin:");
    int source1 = int.parse(stdin.readLineSync());

    // Sample input graph
    // 0 - 1 - 2
    // |     /
    // 3 - 4

    print("Vertices in Depth First Search of the graph in sample input is: ");

    // Depth First Search of the above graph
    Depth_First_Search(graph,source1,no_vertices);

    // Sample Output
    // Vertices in Depth First Search of the graph in sample input is:
    // 0 3 4 2 1

}

File: Depth_First_Search/Depth_First_Search.cs
/// C# program for Depth First Search Traversal

using System; 
using System.Collections.Generic; //used for including List 
  
class DirectedGraph 
{ 
    List<int>[] adjacencylist; 
    int vertix; 
   
    DirectedGraph(int newv) //Constructor with same name as that of class
    { 
        vertix = newv; 
        adjacencylist = new List<int>[newv]; 
        for (int i = 0; i < newv; ++i) 
            adjacencylist[i] = new List<int>(); 
    } 
    
    void DFSFunction(int newv, bool[] traversed) 
    { 
        traversed[newv] = true; 
        Console.Write(newv + " "); 
        List<int> vList = adjacencylist[newv]; 
        foreach (var n in vList) 
        { 
            if (!traversed[n]) 
                DFSFunction(n, traversed); 
        } 
    } 

    void Addition_of_Edge(int newv, int var) 
    { 
        adjacencylist[newv].Add(var); 
    } 
   
    void Depth_First_Search(int newv) 
    { 
        bool[] traversed = new bool[vertix]; 
  
        DFSFunction(newv, traversed); 
    } 
    
  public static void Main(String[] args) 
    { 
        DirectedGraph newgraph = new DirectedGraph(4); // Number of vertices is 4
        Console.WriteLine("Number of edges between the vertices to enter");
        int n = Convert.ToInt32(Console.ReadLine());
        for (int i=0;i<n;i++){
          int x = Convert.ToInt32(Console.ReadLine());
          int y = Convert.ToInt32(Console.ReadLine());
          newgraph.Addition_of_Edge(x, y);
        }
  
        Console.WriteLine("Enter vertex which is the starting point");
        int s = Convert.ToInt32(Console.ReadLine());
        Console.WriteLine("Depth First Search Traversal starting from vertex " + s + " is : "); 
        newgraph.Depth_First_Search(s); 
    } 
} 
// Sample input:
// Number of edges between the vertices to enter
// 6
// 0 2 0 1 2 1 2 0 3 2 2 2
// Enter vertex which is the starting point
// 3

// Sample OP:
// Depth First Search Traversal starting from vertex 3 is :
// 3 2 1 0
